ledger-core/inc/core/utils/AbstractFactoryGenerator.hpp

```diff
@@ -0,0 +1,69 @@
+/*
+ *
+ * AbstractFactoryGenerator
+ * ledger-core
+ *
+ * Created by Alexis Le Provost on 16/09/2019.
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Ledger
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#pragma once 
+
+#include <map>
+#include <memory>
+#include <type_traits>
+
+namespace ledger
+{
+namespace core
+{
+
+template <typename Key, typename T>
+class AbstractFactoryGenerator 
+{
+public:
+    static_assert(
+        std::is_base_of<std::enable_shared_from_this<T>, T>::value,
+        "AbstractFactoryGenerator<Key, T> requires 'T' to be sharable."
+    );
+
+    using key_type = std::decay_t<Key>;
+    using value_type = std::shared_ptr<T>;
+
+    AbstractFactoryGenerator() = default;
+    ~AbstractFactoryGenerator() = default;
+
+    value_type add(key_type const& key, value_type const& factory);
+    void remove(key_type const& key);
+    value_type make(key_type const& key) const;
+
+private:
+    std::map<key_type, value_type> factories_;
```

Comment:
Why not just `unordered_map<string, shared_ptr<AbstractFactory>>`.
There is to many names and abstractions to remember by I already know what is `unordered_map` and how does it behave.

Reply:
This is the current implementation in `WalletPool` - `std::vector` is used but anyway - and I'm ok with that. 
However add a layer of abstraction with a well-known class is always a best practice. The readability is increased and all the burden about the STL containers is hidden.
For example: 
```
auto fIt = std::find_if(_factories.begin(), _factories.end(), [currencyName] (const std::shared_ptr<AbstractWalletFactory>& f) {
                    return currencyName == f->getCurrency().name;
                });
...
                if (fIt != _factories.end())
                    _factories.erase(fIt);
```
compare to
```
factories_.forget(currencyName)
```

Reply:
`However add a layer of abstraction with a well-known class is always a best practice` this is debatable, personally I think this add complexity for relatively small profit, I think we need someones else opinion on this @phaazon @pollastri-pierre 
Your example code is for `vector` for unordered_map it should be much simpler IMO:
```
map.erase(currencyName);
```

Reply:
As I said before, I agree with the current implementation but with the usage of `map` instead  of `vector`. 
In our case, the generator is quite simple - then complexity keeps thin - for now and should probably evolve with the incoming cryptocurrencies.

Reply:
My opinion on this is that we should be using the right data structure to do a given job. If we don’t care about the order, we should be using `unordered_map`. The reason for this is both correctness and performance: maintaining a `std::map`’s internal structure (balancing) to preserve order has a cost. We shouldn’t pay for it if we don’t care.

To give more hindsight, C++ has the naming wrong. `std::map` is okay but `std::unordered_map` should be called `std::hash_map`. Here, a _hash map_ is the right tool so I agree with @teams2ua. For factories I see zero reason why we would need a binary tree map or red-black tree or whatever.

@teams2ua mentioned another point we haven’t talked and that I think we should talk to:

```cpp
     std::map<key_type, value_type> factories_; 
```

`key_type` and `value_type`, here, don’t conform to the naming convention we’ve been using in the codebase, which is PascalCase for types and type variables.

And about `std::vector`, I’m totally fine to use it _first_. We’re talking about factories, which we won’t have a lot of. I think the debate about `std::vector` is sterile as the performance between a linear search and a binary search with that few items will be very small.

Reply:
We all agree to use `std::unordered_map`. The opinion conflict here is to use directly STL container in the WalletPool or use that class I made to abstract the generation of abstract factories.
@phaazon I ignore we have such naming convention about using aliased type - I don't see any in the code base

Reply:
We don’t use aliased types. But types follow the PascalCase convention. :)